src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIService.java

```diff
@@ -0,0 +1,192 @@
+            boolean targetProjectExists) {
+        // No build plans exist for local CI. Only return a plan name.
+        final String cleanPlanName = getCleanPlanName(targetPlanName);
+        return targetProjectKey + "-" + cleanPlanName;
+    }
+
+    @Override
+    public void givePlanPermissions(ProgrammingExercise programmingExercise, String planName) {
+        // Not implemented for local CI.
+    }
+
+    @Override
+    public void removeAllDefaultProjectPermissions(String projectKey) {
+        // Not implemented for local CI.
+    }
+
+    @Override
+    public void giveProjectPermissions(String projectKey, List<String> groupNames, List<CIPermission> permissions) {
+        // Not implemented for local CI.
+    }
+
+    @Override
+    public void enablePlan(String projectKey, String planKey) throws LocalCIException {
+        // Not implemented for local CI. No plans exist that must be enabled.
+    }
+
+    @Override
+    public void updatePlanRepository(String buildProjectKey, String buildPlanKey, String ciRepoName, String repoProjectKey, String newRepoUrl, String existingRepoUrl,
+            String newBranch, List<String> triggeredByRepositories) throws LocalCIException {
+        // Not implemented for local CI. No build plans exist.
+        // When a student pushes to a repository, a build is triggered using the information contained in the participation which includes the relevant repository.
+    }
+
+    /**
+     * Extract the plan key from the Bamboo requestBody.
+     *
+     * @param requestBody The request Body received from the CI-Server.
+     * @return the plan key or null if it can't be found.
+     * @throws BambooException is thrown on casting errors.
+     */
+    @Override
+    public String getPlanKey(Object requestBody) throws LocalCIException {
+        // This method is called in the processNewProgrammingExerciseResult method of the ResultResource.
+        // It is thus never called for local CI as the local CI results directly go to the method processNewProgrammingExerciseResult in the ProgrammingExerciseGradingService.
+        return null;
+    }
+
+    @Override
+    public ConnectorHealth health() {
+        return new ConnectorHealth(true);
+    }
+
+    @Override
+    public Optional<String> getWebHookUrl(String projectKey, String buildPlanId) {
+        // No webhooks needed between local CI and local VC, so we return an empty Optional.
+        return Optional.empty();
+    }
+
+    /**
+     * Gets the latest available artifact for the given participation.
+     *
+     * @param participation to use its buildPlanId to find the artifact.
+     * @return the html representation of the artifact page.
+     */
+    public ResponseEntity<byte[]> retrieveLatestArtifact(ProgrammingExerciseParticipation participation) {
+        // Build artifacts are not supported by local CI yet.
+        log.error("Unsupported action: LocalCIService.retrieveLatestArtifact()");
+        return null;
```

Comment:
Can we return a response entity with an empty array instead of null or do we return null here on purpose?

Reply:
There's no specific reason to return null. I changed it to a 204 response with a 0 length byte array. This method will never be called as build artifacts are not supported by local CI but it's cleaner this way.